balancer-js/src/sdk/index.ts

```diff
@@ -1,22 +1,66 @@
-import { ConfigSdk } from '../types';
-import { Network } from '../constants/network';
+import {
+    BalancerNetworkConfig,
+    BalancerSdkConfig,
+    BalancerSdkSorConfig,
+} from '../types';
 import { SwapsService } from '../swapsService';
 import { RelayerService } from '../relayerService';
+import { SOR } from '@balancer-labs/sor';
+import { SorFactory } from '../sor/sorFactory';
+import { BALANCER_NETWORK_CONFIG } from '../constants/contracts';
+import { JsonRpcProvider, Provider } from '@ethersproject/providers';
+import { createSubgraphClient, SubgraphClient } from '../subgraph/subgraph';
 
 export class BalancerSDK {
-    network: Network;
-    rpcUrl: string;
-    swaps: SwapsService;
-    relayer: RelayerService;
+    public readonly network: BalancerNetworkConfig;
+    public readonly rpcUrl: string;
+    public readonly swaps: SwapsService;
+    public readonly relayer: RelayerService;
+    public readonly sor: SOR;
+    public readonly provider: Provider;
+    public readonly subgraphClient: SubgraphClient;
 
-    constructor(config: ConfigSdk, swapService = SwapsService, relayerService = RelayerService) {
-        this.network = config.network;
+    constructor(config: BalancerSdkConfig) {
+        this.network = this.getNetworkConfig(config);
         this.rpcUrl = config.rpcUrl;
-        this.swaps = new swapService({
-            network: this.network,
-            rpcUrl: this.rpcUrl,
-            subgraphUrl: config.subgraphUrl
-        });
-        this.relayer = new relayerService(this.swaps, this.rpcUrl);
+        this.provider = config.provider ?? new JsonRpcProvider(this.rpcUrl);
+        this.subgraphClient = createSubgraphClient(this.network.subgraphUrl);
+
+        const sorConfig = this.getSorConfig(config);
+        this.sor = SorFactory.createSor(
+            this.network,
+            sorConfig,
+            this.provider,
+            this.subgraphClient
+        );
+
+        this.swaps = new SwapsService(this.network, this.sor, this.provider);
+        this.relayer = new RelayerService(this.swaps, this.rpcUrl);
+    }
+
+    private getNetworkConfig(config: BalancerSdkConfig): BalancerNetworkConfig {
+        if (typeof config.network === 'number') {
+            const networkConfig = BALANCER_NETWORK_CONFIG[config.network];
+
+            return {
+                ...networkConfig,
+                subgraphUrl:
+                    config.customSubgraphUrl ?? networkConfig.subgraphUrl,
+            };
+        }
+
+        return {
+            ...config.network,
+            subgraphUrl: config.customSubgraphUrl ?? config.network.subgraphUrl,
+        };
+    }
+
+    private getSorConfig(config: BalancerSdkConfig): BalancerSdkSorConfig {
+        return {
+            tokenPriceService: 'coingecko',
+            poolDataService: 'subgraph',
+            fetchOnChainBalances: true,
+            ...config.sor,
+        };
```

Comment:
imo we should avoid as much as possible bloating the initial SDK class. Both of these functions should probably come under submodules so they can be called like:
```ts
const balancer = new BalancerSDK(...)

balancer.network.config
balancer.sor.config
```
I also think it's kind of weird that we have to pass in the `BalancerSdkConfig` to these functions. Since they are class functions they should probably have access to the config already. The constructor could instead look something like this:
```ts
class BalancerSDK {
    constructor(public readonly config: BalancerSdkConfig) {
       // Instantiate sub modules
      this.network = new NetworkService(this)
      ...
    }
}

class NetworkService {
    constructor(private readonly sdk: BalancerSDK) {}

    public get config(): BalancerNetworkConfig {
        // Here you have access to the sdk config input via:
       // this.sdk.config
    }
}

// Now you can access the network config like:
balancerSDK.network.config
```
And you could do something similar for the SOR.

Reply:
I'd have to disagree pretty strongly here. It's a point of separation of concerns and creating services that know only what they need to in order to minimize the testable surface area.

If the SDK relies on the SOR and the SOR relies on the SDK, it creates a circular dependency. This should be avoided at all costs imo. 

The `NetworkService` should not know anything about the `BalancerSdk`. It does make access easier, but it creates a dependency structure that will become a nightmare in the long run.

Reply:
Sorry @danielmkm, having re-reviewed the code, I mistook what these functions were for. I thought they were for external access and missed the `private` key and usage. Also, totally agree that this pattern wouldn't work here for the SOR. I still would rather not clutter the initial SDK class as much as possible but for now I think these changes are ok.

One minor thing that I'm not sure about is the fact that `BalancerSdkConfig.network` can be a string | BalancerNetworkConfig. I'm not sure where this came from but would be nice to simplify that if possible. Although, this doesn't need to be addressed in this PR.

Reply:
Hey @garethfuller, thanks for the response!

I see your point on the initial SDK class, I'll refactor it a bit to try to keep it very minimal.

As for the config, I tried to explain the motivation in #1. The goal is trying to find a happy medium between ease of use and customizability from the perspective of the implementer. Because of the flexibility of typescript typing, we can let the implementer simply pass in a network id (ie `Network. MAINNET`), or allow them to fully customize the definition.

The biggest win here is that we allow for customization without having to edit the SDK code itself. Imagine a scenario where you deploy a new `multicaller` or `staBal3Pool` to kovan, and would like to test it quickly. With this setup, you can pass something like the following into the SDK:

```
new BalancerSdk({
    network: {
        ...BALANCER_NETWORK_CONFIG[Network.KOVAN],
        multicall: 'new muticall address',
        staBal3Pool: 'new stabal3 config'
    },
    rpcUrl
});
```

The "normal" case allows the implementer to simply pass in the network id, without worrying about anything underlying

```
const balancerSdk = new BalancerSdk({ network: Network.MAINNET, rpcUrl});
``` 

Another scenario is for beethoven on fantom. The flexible definition allows us to pass in all of our contracts to the `BalancerSdk` constructor, it eliminates the need for us to immediately fork the repo just to use it.

Reply:
I move the `getSorConfig` function into the `SorFactory` to thin out the `BalancerSdk` class. I'm not sure the network needs its own service as its more config than anything. Left it for the moment, but if you have a strong opinion on it I can move it as well.

https://github.com/balancer-labs/balancer-sdk/pull/22/commits/fb68f672133cebb61a85e8c6140e06df7c4a688a
https://github.com/balancer-labs/balancer-sdk/pull/22